regparser/api_writer.py

```diff
@@ -123,16 +131,396 @@ def write(self, python_object):
+                and root.label[1] in self.caps:
+            # this is the case for hyphenated appendices like MS-1 in reg X
+            elem = Element('interpAppSection', appendixLetter=root.label[1], label=root.label_id())
+            title = SubElement(elem, 'title')
+            title.text = root.title
+        elif root.node_type == 'interp': # and \
+                #((len(root.label) > 3 and root.label[-1] != 'Interp') or \
+                # (len(root.label) == 3 and root.label[1] == 'Interp')):
+            # fall-through for all other interp nodes, which should be paragraphs
+            target = [item for item in root.label if item != 'Interp']
+            target = '-'.join(target)
+            label = root.label_id()
+            elem = Element('interpParagraph', label=label, target=target)
+            if root.title:
+                title = SubElement(elem, 'title')
+                title.text = root.title
+            text = self.apply_layers(root)
+            content = fromstring('<content>' + text + '</content>')
+            elem.append(content)
+        else:
+            elem = Element('paragraph', label=root.label_id(), marker=root.label[-1])
+            if root.title:
+                title = SubElement(elem, 'title')
+                title.text = root.title
+            text = self.apply_layers(root)
+            content = fromstring('<content>' + text + '</content>')
+            elem.append(content)
+
+        if len(root.label) > 1 and ('Subpart' not in root.label):
+            # the part is a special case
+            for child in root.children:
+                sub_elem = self.to_xml(child)
+                elem.append(sub_elem)
+        return elem
+
+    def apply_layers(self, node):
+        text = node.text
+        replacement_hashes = {}
+        all_offsets = []
+        all_replacements = []
+        # import pdb; pdb.set_trace()
+        for ident, layer in self.layers.items():
+            # print node.label_id()
+            if node.label_id() in layer:
+                # print 'applying layers'
+                replacements = layer[node.label_id()]
+                if ident == 'terms':
+                    offsets, repls = XMLWriteContent.apply_terms(text, replacements)
+                elif ident == 'paragraph-markers':
+                    offsets, repls = XMLWriteContent.apply_paragraph_markers(text, replacements)
+                elif ident == 'internal-citations':
+                    offsets, repls = XMLWriteContent.apply_internal_citations(text, replacements)
+                elif ident == 'definitions':
+                    offsets, repls = XMLWriteContent.apply_definitions(text, replacements)
+                elif ident == 'external-citations':
+                    offsets, repls = XMLWriteContent.apply_external_citations(text, replacements)
+                else:
+                    offsets = []
+                    repls = []
+
```

Comment:
Yes, so! From above, I think these should be `apply()` methods on the layers themselves.